securedrop/source_app/main.py

```diff
@@ -104,7 +104,14 @@ def lookup():
     @login_required
     def submit():
         msg = request.form['msg']
-        fh = request.files['fh']
+        files = request.files
+        fh = files.getlist('fh[]')
+
+        #  Based on http://flask.pocoo.org/docs/0.12/patterns/fileuploads/
+        #  if user does not select file, browser also
+        #  submit a empty part without filename
+        if fh and not fh[0].filename:
```

Comment:
So since `fh` is a list of `FileStorage` objects we can just enforce the max number of files here in Python using the length of `fh` (e.g. `if len(fh) < NUMBER_OF_MAX_FILES` then we proceed with the upload, else we flash a message "You can only upload up to 10 files" or similar). Once that's in place it should address the concern @emkll and I were discussing in this PR and we should be good to go to merge this in! What do you think?

Reply:
That sounds _great_—a fine compromise and requires no JS. Please add! Happy to re-review by hammering away with submissions to confirm we have a limit. 

UX consideration: we may want to inform the uploader about the file max _before_ uploading—but happy to test before commenting more about design implications.

Reply:
I idea is that because fh is a list in the server side, means we already have all the data in the memory (or open file descriptors). So, forcing on the list, means we are doing it after everything is on the server.

Reply:
Do note that with the default Werkzeug `stream_factory`, the files will have be attempted to be generated/opened at the moment you call the first `request.form` or `request.files` (https://github.com/pallets/werkzeug/blob/master/werkzeug/formparser.py#L45).

So what you would need to do is override `werkzeug.wrapper.BaseRequest._get_file_stream` (which you have access to as `flask.request._get_file_stream` I think), to return a different stream_factory (maybe always use BytesIO, or do something smarter like "count and abort" in the stream factory), and this must be setup before `_load_form_data` gets called (i.e. before you access `request.form` or `request.files`).

All this is to work around the fact that I don't see any limiting code in https://github.com/pallets/werkzeug/blob/master/werkzeug/formparser.py#L488.
I think that ideally, Werkzeug/Flask would check the max number of files opened.

(Note that this is just from reading the Werkzeug sources. I have not actually tried this)

Reply:
I can actually achieve this easily (say take only first 10 files and discard rest) if I directly patch the `formparser.py` (tested on my development vm). I will try to figure out a good way to money patch at the runtime.

Reply:
I have a patch ready for monkey patching which takes the first 10 files in the request and discards the rest. Will update the PR after I manage to test locally in stage vms.